### PR TITLE
Fuse ops in eager cosine_similarity while keeping the stability and the gradients

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104772
* __->__ #104771

There was a regression in https://github.com/pytorch/pytorch/pull/31378
which was reported in https://github.com/pytorch/pytorch/issues/104564.
This PR should keep the efficiency and memory usage from the original
implementation, while keeping the stability of the latter.

This solution was already discussed in https://github.com/pytorch/pytorch/pull/31378,
but it was discarded because it modified the vector_norm in-place. The
only magic ingredient that was missing for that solution to work was to
add a `clone()` after calling the `vector_norm`.

I hope this PR takes shorter to land than https://github.com/pytorch/pytorch/issues/104564.

Fixes https://github.com/pytorch/pytorch/issues/104564

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Comment by lezcano** [comment]
This implements on eager the solution discussed in https://github.com/pytorch/pytorch/issues/104564#issuecomment-1625348908

File: aten/src/ATen/native/Distance.cpp [hunk 1]
```diff
@@ -271,71 +272,31 @@ Tensor _pdist_backward(const Tensor& grad, const Tensor& self, const double p, c
...
-   */
+
   auto commonDtype = at::result_type(x1_, x2_);
   TORCH_CHECK(at::isFloatingType(commonDtype), "expected common dtype to be floating point, yet common dtype is ", commonDtype);
 
-  auto common_size = at::infer_size_symdimvector(x1_.sym_sizes(), x2_.sym_sizes());
-  auto x1 = x1_.to(commonDtype).expand_symint(common_size);
-  auto x2 = x2_.to(commonDtype).expand_symint(common_size);
+  // We accept integral types but vector_norm does not
+  auto x1_int = isIntegralType(x1_.scalar_type());
```
Comment (albanD) [inline aten/src/ATen/native/Distance.cpp:280]: nit: `x1_is_int`

File: aten/src/ATen/native/Distance.cpp [hunk 2]
```diff
@@ -271,71 +272,31 @@ Tensor _pdist_backward(const Tensor& grad, const Tensor& self, const double p, c
 }
 
 Tensor cosine_similarity(const Tensor& x1_, const Tensor& x2_, int64_t dim, double eps) {
-  /*
-   * cosine_similarity(x1, x2) = <x1, x2> / (||x1|| * ||x2||)
-   *
-   * The current implementation is an improvement over the previous version.
-   *
-   * Previous implementation:
```
Comment (albanD) [inline aten/src/ATen/native/Distance.cpp:279]: I think the comparison with the old implementation is still relevant in terms of numerical precision.
Also this comment is still valid for the updated implementation right?

Comment (lezcano) [reply aten/src/ATen/native/Distance.cpp:279]: reverted

File: aten/src/ATen/native/Distance.cpp (hunk 1 above)
Comment (lezcano) [reply aten/src/ATen/native/Distance.cpp:280]: done

**Comment by lezcano** [comment]
@albanD addressed the comments

**Review by albanD — APPROVED** [review]
SGTM

**Pull request closed without merging**